File: benchmarks/successor/representation-v1/transformed/descriptive/csharp/gold/002-overdue/Program.cs
using System.Text.Json;

namespace OrderFlow;

public sealed class Order
{
    public string Id { get; init; } = "";
    public DateTimeOffset CreatedAt { get; init; }
    public string Status { get; init; } = "";
    public int? Priority { get; init; }
    public DateTimeOffset? DueAt { get; init; }
}

public sealed class Request
{
    public string Operation { get; init; } = "";
    public Order[] Orders { get; init; } = [];
    public DateTimeOffset? AsOf { get; init; }
}

public sealed record Response(string[] Ids);

public static class Program
{
    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    public static void Main()
    {
        string? line;
        while ((line = Console.ReadLine()) is not null)
        {
            try
            {
                var request = JsonSerializer.Deserialize<Request>(line, Options)
                    ?? throw new InvalidOperationException("Request was null");
                var response = Handle(request);
                Console.WriteLine(JsonSerializer.Serialize(response, Options));
            }
            catch (Exception ex)
            {
                Console.WriteLine(JsonSerializer.Serialize(new { error = ex.Message }, Options));
            }
        }
    }

    private static Response Handle(Request request)
    {
        IEnumerable<Order> selected;
        if (string.Equals(request.Operation, "ready", StringComparison.OrdinalIgnoreCase))
        {
            selected = request.Orders
                .Where(order => string.Equals(order.Status, "pending", StringComparison.OrdinalIgnoreCase))
                .OrderByDescending(order => order.Priority ?? 0)
                .ThenBy(order => order.CreatedAt)
                .ThenBy(order => order.Id, StringComparer.Ordinal);
        }
        else if (string.Equals(request.Operation, "overdue", StringComparison.OrdinalIgnoreCase))
        {
            var asOf = request.AsOf
                ?? throw new ArgumentException("asOf is required for overdue", nameof(request.AsOf));
            selected = request.Orders
                .Where(order =>
                    (string.Equals(order.Status, "pending", StringComparison.OrdinalIgnoreCase)
                     || string.Equals(order.Status, "processing", StringComparison.OrdinalIgnoreCase))
                    && order.DueAt.HasValue
                    && order.DueAt.Value < asOf)
                .OrderBy(order => order.DueAt!.Value)
                .ThenByDescending(order => order.Priority ?? 0)
                .ThenBy(order => order.Id, StringComparer.Ordinal);
        }
        else
        {
            throw new ArgumentException($"Unknown operation: {request.Operation}", nameof(request.Operation));
        }

        return new Response(selected.Select(order => order.Id).ToArray());
    }
}

File: benchmarks/successor/representation-v1/transformed/descriptive/fsharp/gold/002-overdue/Program.fs
open System
open System.Text.Json

[<CLIMutable>]
type Order =
    { id: string
      createdAt: DateTimeOffset
      status: string
      priority: Nullable<int>
      dueAt: Nullable<DateTimeOffset> }

[<CLIMutable>]
type Request =
    { operation: string
      orders: Order array
      asOf: Nullable<DateTimeOffset> }

[<CLIMutable>]
type Response = { ids: string array }

let options =
    JsonSerializerOptions(
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    )

let isStatus expected actual =
    String.Equals(expected, actual, StringComparison.OrdinalIgnoreCase)

let priorityOf (order: Order) =
    if order.priority.HasValue then order.priority.Value else 0

let ready (orders: Order array) =
    orders
    |> Array.filter (fun order -> isStatus "pending" order.status)
    |> Array.sortBy (fun order -> -(priorityOf order), order.createdAt, order.id)

let overdue asOf (orders: Order array) =
    orders
    |> Array.filter (fun order ->
        (isStatus "pending" order.status || isStatus "processing" order.status)
        && order.dueAt.HasValue
        && order.dueAt.Value < asOf)
    |> Array.sortBy (fun order -> order.dueAt.Value, -(priorityOf order), order.id)

let handle (request: Request) =
    let selected =
        match request.operation with
        | operation when isStatus "ready" operation -> ready request.orders
        | operation when isStatus "overdue" operation ->
            if not request.asOf.HasValue then
                invalidArg "asOf" "asOf is required for overdue"
            overdue request.asOf.Value request.orders
        | operation -> invalidArg "operation" $"Unknown operation: {operation}"

    selected |> Array.map (fun order -> order.id) |> fun ids -> { ids = ids }

let mutable running = true

while running do
    match Console.ReadLine() with
    | null -> running <- false
    | input ->
      try
        let request = JsonSerializer.Deserialize<Request>(input, options)
        match request with
        | null -> raise (InvalidOperationException("Request was null"))
        | request ->
          let response = handle request
          Console.WriteLine(JsonSerializer.Serialize(response, options))
      with ex ->
        Console.WriteLine(JsonSerializer.Serialize({| error = ex.Message |}, options))
